matlab/CMakeLists.txt

```diff
@@ -72,3 +89,28 @@ else()
   set_target_properties(featherwritemex PROPERTIES LIBRARY_OUTPUT_DIRECTORY
                                                    $<1:${CMAKE_SOURCE_DIR}/src>)
 endif()
+
+# ######################
+# GoogleTest Integration
+# ######################
+# If the user has specified a GTEST_ROOT value, use their pre-built GoogleTest binaries.
+# Otherwise, download and build GoogleTest automatically.
+if(GTEST_ROOT)
+  find_package(GTest REQUIRED)
+else()
+  download_and_build_google_test()
```

Comment:
Do we need this functionality for MATLAB module?
How about installing GoogleTest via package manager on system such as `apt`, `dnf`, `brew` and so on?

Reply:
OK.
How about referring GoogleTest bundled in Apache Arrow C++ instead of adding this functionality to the MATLAB library?
If we build bundled GoogleTest in Apache Arrow C++, it's installed into `${ARROW_BUILD_DIR}/googletest_ep-prefix/{include,lib}/`. We can find it with `find_package(GTest REQUIRED CONFIG)` and `-DGTest_ROOT=${ARROW_BUILD_DIR}/googletest_ep-prefix`.
And `find_package(GTest REQUIED COFNIG)` without `-DGTest_ROOT` will find GoogleTest installed into system by package manager.

Reply:
Setting `GTEST_ROOT` to `${ARROW_BUILD_DIR}/googletest_ep-prefix` inside the `CMakeLists.txt` for MATLAB seems like a reasonable approach.

If we could automatically trigger a build of the Arrow C++ libraries with `ARROW_BUILD_TESTS=ON` (in order to get GoogleTest to be built in `${ARROW_BUILD_DIR}/googletest_ep-prefix`) from the MATLAB `CMakeLists.txt` file, that would be great. But, I'm not sure what the best approach is to accomplish this.

It seems like it would be error prone to expect users to manually build the Arrow C++ libraries first due to the variation in names they might use for their CMake build folder, as well as the different flags they might pass to the build (e.g. they might forget to include `-DARROW_BUILD_TESTS=on`).

Reply:
We tried invoking an automatic build of the Arrow C++ libraries from `matlab/CMakeLists.txt` using:

```cmake
ExternalProject_Add(ArrowCpp SOURCE_DIR "${CMAKE_SOURCE_DIR}/../cpp" CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=. -DARROW_BUILD_TESTS=ON)
```

But, unfortunately, this doesn't seem to be exporting targets for `arrow_shared` or `GTest::GTest`. Admittedly, I am not particularly familiar with `ExternalProject_Add`.

Is this the approach you were suggesting? Or, were you thinking that users would just manually build the Arrow C++ libraries first, in order to get access to the bundled GoogleTest binares (i.e. similar to https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst#build-and-test)?

Reply:
Can we clarify target people of the MATLAB library's test?

1. Users: People who use the MATLAB library.
2. Developers: People who develop the MATLAB library.
3. Both of 1. and 2.

My thought is 2.

For 2., custom Apache Arrow C++ build is useful. For example, developers may want to build with `-DCMAKE_BUILD_TYPE=debug` for debugging and with `-DCMAKE_BUILD_TYPE=release` for benchmarking.

So, it may be reasonable that developers build Apache Arrow C++ manually.

For 1., users may not want to enable tests by default. Because they just want to use the MATLAB library.

If you still want to implement automatic Apache Arrow C++ build in CMake, I'll take a look.

Reply:
Thanks for sharing your thoughts, @kou!

-------------

> Can we clarify target people of the MATLAB library's test?

I agree with you that _Developers_ are the primary target audience of the C++ tests for the MATLAB interface.

> For 2., custom Apache Arrow C++ build is useful. For example, developers may want to build with -DCMAKE_BUILD_TYPE=debug for debugging and with -DCMAKE_BUILD_TYPE=release for benchmarking

You make an excellent point about providing developers with the flexibility to build their own custom `libarrow` for debugging and benchmarking purposes! This is definitely something that we would like to support. If we use `find_package`, as you previously suggested, this should support custom `libarrow` builds via `ARROW_HOME`, as well as package managers.

> For 1., users may not want to enable tests by default. Because they just want to use the MATLAB library.

Good point! I think it would make sense to introduce a flag like `MATLAB_BUILD_TESTS`, with a default value of `OFF`, so that users can toggle building of the C++ tests for the MATLAB interface.

-------------

Based on your input, we went back to the drawing board and learned more about `ExternalProject_Add` and `IMPORTED` targets. By studying the source code for `build_dependency` in `ThirdpartyToolchain` more closely, I now understand how `ExternalProject_Add` can be used to trigger an automatic build of the Apache Arrow C++ libraries (and bundled GoogleTest).

I put together a prototype which automatically builds Arrow C++ and bundled GoogleTest on Linux (only if the user has specified `-DMATLAB_BUILD_TESTS=ON`).

**You can find the prototype here:**

https://github.com/mathworks/arrow/commit/42609d9671501c8fe1e8c613337eedb207d32e91

I tested this manually under the following conditions:

1. GoogleTest installed via the `apt` package manager to `/usr/lib/...` on Debian 10
2. A custom `GTEST_ROOT` and a custom `ARROW_HOME` value are specified
3. `find_package` fails to find a valid Arrow or GTest installation via 1. or 2., and an automatic C++ source build is triggered

The code could probably benefit from some additional refactoring, and I'm not sure whether "Modern CMake Best Practices" would deem it acceptable to depend on the bundled GoogleTest binaries in this way, but it does seem to be working as expected.

Let me know what you think. 

Thanks!

Reply:
Could you merge the commit to this branch?
Then I can comment on the implementation in this pull request.

Reply:
@kou - my apologies for the delay! I was away for a few days.

I've pushed the changes from mathworks@42609d9 to this branch, so that you can comment on the implementation.

Thank you!